runtime/src/accounts_cache.rs

```diff
@@ -147,14 +147,22 @@ impl AccountsCache {
     }
 
     pub fn add_root(&self, root: Slot) {
-        self.maybe_unflushed_roots.write().unwrap().insert(root);
+        let max_flushed_root = self.fetch_max_flush_root();
+        if root > max_flushed_root || (root == max_flushed_root && root == 0) {
+            self.maybe_unflushed_roots.write().unwrap().insert(root);
+        }
     }
 
-    pub fn clear_roots(&self) -> HashSet<Slot> {
-        std::mem::replace(
-            &mut self.maybe_unflushed_roots.write().unwrap(),
-            HashSet::new(),
-        )
+    pub fn clear_roots(&self, max_root: Option<Slot>) -> BTreeSet<Slot> {
+        let mut w_maybe_unflushed_roots = self.maybe_unflushed_roots.write().unwrap();
+        if let Some(max_root) = max_root {
+            let greater_than_max_root = w_maybe_unflushed_roots.split_off(&(max_root + 1));
+
+            // After the swap, `greater_than_max_root` now contains slots <= `max_root`
```

Comment:
well, my comment was confusing. I meant this comment can just be removed. `std::mem::replace` takes ownership of the 2nd argument. So, `greater_than_max_root` is just gone after the replace.  So current comment a bit is confusing, conflicting with the rust mental model. Considering the similarity with the `else` code, I think there is no need to comment about this anymore. Thanks to your cleaning, the code is now eloquent by itself. :)

I mean, in both cases of `if`, the code itself just fluently reads like this (assuming reader knows about `std::mem::replace`):

> `w_maybe_unflushed_roots`'s content is replaced with the 2nd argument, taking ownership, returning its old content

If you wanted a comment how about this?:

```rust
// After the replace, w_maybe_unflushed_roots contains slots <= `max_root`
```

or 

```rust
// After the replace, `w_maybe_unflushed_roots` (replaced with `greater_than_max_root`) now contains slots <= `max_root`
```

Reply:
Ah yup, forgot that I replaced the `swap)_` with `replace()`. Updated, I think you meant:

```rust
// After the replace, w_maybe_unflushed_roots contains slots > `max_root`
```